airbyte-commons/src/main/java/io/airbyte/commons/json/JsonPaths.java

```diff
@@ -0,0 +1,278 @@
+   * @param replacement - a string value to replace the current value at the jsonPath
+   */
+  public static JsonNode replaceAtString(final JsonNode json, final String jsonPath, final String replacement) {
+    return replaceAtJsonNode(json, jsonPath, Jsons.jsonNode(replacement));
+  }
+
+  /**
+   * Traverses into a json object and replaces all values that match the input path with the provided
+   * json object. Does nothing if no existing fields match the path.
+   *
+   * @param json - json object
+   * @param jsonPath - path into the json object. must be in the format of JSONPath.
+   * @param replacement - a json node to replace the current value at the jsonPath
+   */
+  public static JsonNode replaceAtJsonNodeLoud(final JsonNode json, final String jsonPath, final JsonNode replacement) {
+    assertIsJsonPath(jsonPath);
+    return JsonPath.parse(Jsons.clone(json)).set(jsonPath, replacement).json();
+  }
+
+  /**
+   * Traverses into a json object and replaces all values that match the input path with the provided
+   * json object. Does nothing if no existing fields match the path.
+   *
+   * @param json - json object
+   * @param jsonPath - path into the json object. must be in the format of JSONPath.
+   * @param replacement - a json node to replace the current value at the jsonPath
+   */
+  public static JsonNode replaceAtJsonNode(final JsonNode json, final String jsonPath, final JsonNode replacement) {
+    try {
+      return replaceAtJsonNodeLoud(json, jsonPath, replacement);
+    } catch (final PathNotFoundException e) {
+      LOGGER.debug("Path not found", e);
+      return Jsons.clone(json); // defensive copy in failure case.
+    }
+  }
+
+  /**
+   * Traverses into a json object and replaces all values that match the input path with the output of
+   * the provided function. Does nothing if no existing fields match the path.
+   *
+   * @param json - json object
+   * @param jsonPath - path into the json object. must be in the format of JSONPath.
+   * @param replacementFunction - a function that takes in a node that matches the path as well as the
+   *        path to the node itself. the return of this function will replace the current node.
+   */
+  public static JsonNode replaceAt(final JsonNode json, final String jsonPath, final BiFunction<JsonNode, String, JsonNode> replacementFunction) {
+    JsonNode clone = Jsons.clone(json);
+    assertIsJsonPath(jsonPath);
+    try {
+      final List<String> foundPaths = getPaths(clone, jsonPath);
+      for (final String foundPath : foundPaths) {
+        final Optional<JsonNode> singleValue = getSingleValue(clone, foundPath);
+        if (singleValue.isPresent()) {
+          final JsonNode replacement = replacementFunction.apply(singleValue.get(), foundPath);
+          clone = replaceAtJsonNode(clone, foundPath, replacement);
+        }
+      }
+    } catch (final PathNotFoundException e) {
+      LOGGER.debug("Path not found", e);
```

Comment:
Is it possible to be in this catch since we are using the `getPaths` method to get all the paths.

Reply:
good point. removed.